File: services/order-placement/internal/core/dto/order.go
package dto

import (
	"time"
)

// OrderItemInput - Input data for creating order items
type OrderItemInput struct {
	ProductID     string `json:"productId" validate:"required"`
	ProductName   string `json:"productName" validate:"required"`
	ProductSku    string `json:"productSku"`
	ImageUrl      string `json:"imageUrl" validate:"required"`
	VariantID     string `json:"variantId"`
	VariantName   string `json:"variantName"`
	OriginalPrice int64  `json:"originalPrice" validate:"required,min=0"`
	SalePrice     int64  `json:"salePrice" validate:"required,min=0"`
	Quantity      int    `json:"quantity" validate:"required,min=1"`
}

// CreateOrderRequest - Request to create a new order
type CreateOrderRequest struct {
	// Shop
	ShopID string `json:"shopId" validate:"required"`

	// Shipping information
	RecipientName   string `json:"recipientName" validate:"required"`
	RecipientPhone  string `json:"recipientPhone" validate:"required"`
	DeliveryAddress string `json:"deliveryAddress" validate:"required"`

	// Payment
	PaymentMethod string `json:"paymentMethod" validate:"required"`

	// Shipping
	ShippingMethod string `json:"shippingMethod"`
	ShippingFee    int64  `json:"shippingFee"`

	// Discount
	Discount map[string]interface{} `json:"discount"`

	// Notes
	CustomerNote string `json:"customerNote"`

	// Items
	Items []OrderItemInput `json:"items" validate:"required,min=1,dive"`
}

// OrderDTO - Data transfer object for order
type OrderDTO struct {
	ID          int64  `json:"id"`
	OrderNumber string `json:"orderNumber"`
	UserID      string `json:"userId"`
	ShopID      string `json:"shopId"`

	// Shipping information
	RecipientName   string `json:"recipientName"`
	RecipientPhone  string `json:"recipientPhone"`
	DeliveryAddress string `json:"deliveryAddress"`

	// Status
	Status        string     `json:"status"`
	PaymentMethod string     `json:"paymentMethod"`
	PaymentStatus string     `json:"paymentStatus"`
	PaidAt        *time.Time `json:"paidAt,omitempty"`

	// Pricing
	Subtotal    int64                  `json:"subtotal"`
	ShippingFee int64                  `json:"shippingFee"`
	Discount    map[string]interface{} `json:"discount"`
	TotalAmount int64                  `json:"totalAmount"`

	// Shipping details
	ShippingMethod    string     `json:"shippingMethod,omitempty"`
	ShippingProvider  string     `json:"shippingProvider,omitempty"`
	TrackingNumber    string     `json:"trackingNumber,omitempty"`
	EstimatedDelivery *time.Time `json:"estimatedDelivery,omitempty"`
	ActualDelivery    *time.Time `json:"actualDelivery,omitempty"`

	// Notes
	CustomerNote string `json:"customerNote,omitempty"`
	SellerNote   string `json:"sellerNote,omitempty"`
	CancelReason string `json:"cancelReason,omitempty"`

	// Timestamps
	ConfirmedAt  *time.Time `json:"confirmedAt,omitempty"`
	ProcessingAt *time.Time `json:"processingAt,omitempty"`
	ShippedAt    *time.Time `json:"shippedAt,omitempty"`
	DeliveredAt  *time.Time `json:"deliveredAt,omitempty"`
	CompletedAt  *time.Time `json:"completedAt,omitempty"`
	CancelledAt  *time.Time `json:"cancelledAt,omitempty"`
	CreatedAt    time.Time  `json:"createdAt"`
	UpdatedAt    time.Time  `json:"updatedAt"`

	OrderItems []OrderItemDTO `json:"orderItems"`
}

// OrderItemDTO - Data transfer object for order item
type OrderItemDTO struct {
	ProductID     string `json:"productId"`
	ProductName   string `json:"productName"`
	ProductSku    string `json:"productSku,omitempty"`
	ImageUrl      string `json:"imageUrl"`
	VariantID     string `json:"variantId,omitempty"`
	VariantName   string `json:"variantName,omitempty"`
	OriginalPrice int64  `json:"originalPrice"`
	SalePrice     int64  `json:"salePrice"`
	Quantity      int    `json:"quantity"`
	Subtotal      int64  `json:"subtotal"`
}

// CreateOrderResponse - Response after creating an order
type CreateOrderResponse struct {
	Order   OrderDTO `json:"order"`
	Message string   `json:"message"`
}

File: services/orders/src/main/java/com/ecom/orders/core/domain/model/Order.java
package com.ecom.orders.core.domain.model;

import com.ecom.orders.infras.adapter.outbound.persistence.converter.JsonBConverter;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Entity
@Table(name = "orders")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Order {

   @Id
   private Long id;

   @Column(name = "order_number", unique = true, nullable = false, length = 50)
   private String orderNumber;

   @Column(name = "user_id", nullable = false)
   private String userId;

   @Column(name = "shop_id", nullable = false)
   private String shopId;

   // Shipping information
   @Column(name = "recipient_name", nullable = false)
   private String recipientName;

   @Column(name = "recipient_phone", nullable = false, length = 20)
   private String recipientPhone;

   @Column(name = "delivery_address", nullable = false, columnDefinition = "TEXT")
   private String deliveryAddress;

   // Order status
   @Column(name = "status", nullable = false, length = 50)
   @Enumerated(EnumType.STRING)
   @Builder.Default
   private OrderStatus status = OrderStatus.UNCONFIRMED;

   // Payment information
   @Column(name = "payment_method", nullable = false, length = 50)
   private String paymentMethod;

   @Column(name = "payment_status", nullable = false, length = 50)
   @Builder.Default
   private String paymentStatus = "UNPAID";

   @Column(name = "paid_at")
   private Instant paidAt;

   // Pricing
   @Column(name = "subtotal", nullable = false)
   @Builder.Default
   private Long subtotal = 0L;

   @Column(name = "shipping_fee", nullable = false)
   @Builder.Default
   private Long shippingFee = 0L;

   @Column(name = "discount", columnDefinition = "jsonb")
   @JdbcTypeCode(SqlTypes.JSON)
   @Builder.Default
   private Map<String, Object> discount = Map.of();

   @Column(name = "total_amount", nullable = false)
   @Builder.Default
   private Long totalAmount = 0L;

   // Shipping
   @Column(name = "shipping_method", length = 50)
   private String shippingMethod;

   @Column(name = "shipping_provider", length = 100)
   private String shippingProvider;

   @Column(name = "tracking_number", length = 100)
   private String trackingNumber;

   @Column(name = "estimated_delivery")
   private Instant estimatedDelivery;

   @Column(name = "actual_delivery")
   private Instant actualDelivery;

   // Notes
   @Column(name = "customer_note", columnDefinition = "TEXT")
   private String customerNote;

   @Column(name = "seller_note", columnDefinition = "TEXT")
   private String sellerNote;

   @Column(name = "cancel_reason", columnDefinition = "TEXT")
   private String cancelReason;

   // Timestamps
   @Column(name = "confirmed_at")
   private Instant confirmedAt;

   @Column(name = "processing_at")
   private Instant processingAt;

   @Column(name = "shipped_at")
   private Instant shippedAt;

   @Column(name = "delivered_at")
   private Instant deliveredAt;

   @Column(name = "completed_at")
   private Instant completedAt;

   @Column(name = "cancelled_at")
   private Instant cancelledAt;

   @Column(name = "created_at", nullable = false, updatable = false)
   @Builder.Default
   private Instant createdAt = Instant.now();

   @Column(name = "updated_at", nullable = false)
   @Builder.Default
   private Instant updatedAt = Instant.now();

   @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
   @Builder.Default
   private List<OrderItem> orderItems = new ArrayList<>();

   @PreUpdate
   public void preUpdate() {
      this.updatedAt = Instant.now();
   }

   // Helper methods for managing bidirectional relationship
   public void addOrderItem(OrderItem orderItem) {
      orderItems.add(orderItem);
      orderItem.setOrder(this);
   }

   public void removeOrderItem(OrderItem orderItem) {
      orderItems.remove(orderItem);
      orderItem.setOrder(null);
   }

   public void calculateTotalAmount() {
      this.subtotal = orderItems.stream()
            .mapToLong(OrderItem::getSubtotal)
            .sum();
      this.totalAmount = this.subtotal + this.shippingFee;
   }
}
